src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs

```diff
@@ -73,7 +73,59 @@ public HostOutputMessage HostOutputMessage
         public async Task<IDelayedException> TryExecuteAsync(IFunctionInstance functionInstance, CancellationToken cancellationToken)
         {
             var functionInstanceEx = (IFunctionInstanceEx)functionInstance;
-            var logger = _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(functionInstance.FunctionDescriptor.LogName));
+            var logger = _loggerFactory.CreateLogger(LogCategories.CreateFunctionCategory(functionInstanceEx.FunctionDescriptor.LogName));
+
+            if (functionInstance.FunctionDescriptor.RetryStrategy != null)
+            {
+                return await ExecuteWithRetriesAsync(functionInstanceEx, cancellationToken, logger, functionInstance.FunctionDescriptor.RetryStrategy);
+            }
+
+            return await TryExecuteAsyncCore(functionInstanceEx, cancellationToken, logger);
+        }
+
+        internal async Task<IDelayedException> ExecuteWithRetriesAsync(IFunctionInstanceEx functionInstance, CancellationToken token, ILogger logger, IRetryStrategy retryStrategy)
+        {
+            var attempt = 0;
+            IDelayedException functionResult = null;
+            TimeSpan nextDelay;
+            do
+            {
+                if (token.IsCancellationRequested)
+                {
+                    logger.LogDebug($"Function execution cancelled.");
+                    break;
+                }
+                logger.LogFunctionRetryAttempt(attempt, retryStrategy.MaxRetryCount);
+                functionResult = await TryExecuteAsyncCore(functionInstance, token, logger);
+                if (functionResult == null)
+                {
+                    // function invocation succeeded.
+                    break;
+                }
+
+                // Build retry context
+                var retryContext = new RetryContext
+                {
+                    RetryCount = ++attempt,
+                    Exception = functionResult.Exception,
+                    Instance = functionInstance
+                };
+
+                if (attempt >= retryStrategy.MaxRetryCount)
+                {
+                    // no.of retries exceeded
+                    break;
+                }
+                nextDelay = retryStrategy.GetNextDelay(retryContext);
+                logger.LogDebug($"Waiting for {nextDelay} before retrying function execution.");
```

Comment:
Some of the logs above use precompiled log functions, but this one doesn't, even though you appear to have declared one for this (LogFunctionRetryDelay). Please make these consistent. For this error path, I don't think precompiled are actually required since we're in an error state (not hot path). Your call. Might be easier to just do direct logging in here.